README.md
# brmcalendar

## 概要

BRMの予定を時系列にカレンダーに並べて表示するだけのスクリプト。

## 説明

To make a code block, indent four spaces:



* 下記のファイルを同じフォルダに置く:

フォルダ構造:

    brmcalendar/
        index.php        -- メインページ
        index.css        -- スタイルシート
        icon.png         -- ページアイコン (iOS用)
        brmdata.php      -- BRMデータ読み込みクラス
        cache.manifest   -- オフラインアプリ用 manifest
        yyyy.txt         -- 西暦４桁のテキストファイル

* yyyy.txtの中身

    - ;で始まる行はコメント行
    - 　ファイル名は yyyy.txt (UTF-8)
    - 中身はcsv形式
       -  ClubUniqueID,クラブ名称,略称名,距離,mmdd,mmdd,...


## LICENSE

このスクリプトはMITライセンスとします。
LICENSE ファイル参照。


## 連絡先

[email :] mmgithub@gmail.com


HISTORY

履歴
=======

Ver 1.0.1
---------
初期バージョン




brmcalendar/brmdata.php
<?php
	define("START_YEAR", 2012);
	$distanceArray = array(200,300,400,600,1000,1200,0);

	class Club {
		var $id;
		var $name;
		var $shortName;
		var $distanceDate;

		function setClub($id, $name, $shortName){
			$this->id = $id;
			$this->name = $name;
			$this->shortName = $shortName;

			$this->distanceDate = array(array(), array(), array(), array(), array(), array());
		}

		function addDate($distance, $date){
			global $distanceArray;

			$i=0;
			while($distanceArray[$i] != 0){
				if($distanceArray[$i] == $distance){
					$this->distanceDate[$i][] = $date;
					break;
				}
				$i++;
			}
		}

		function getBRMData($date, $distanceId, &$BRMDatalist){
			global $distanceArray;

			for($j=0; $j<count($this->distanceDate[$distanceId]); $j++){
				if($this->distanceDate[$distanceId][$j] == $date){
					$BRMDatalist[] = $this->shortName . $distanceArray[$distanceId];
				}
			}
		}
	}

	class BRM {
		var $year;
		var $dateList;
		var $clubList;

		function initialize($year){
			$lastClubId = 0;
			$this->year = $year;
			$this->dateList = array();

			if($handle = fopen($year . ".txt", "r")){
				while(($line = fgetcsv($handle)) !== FALSE){
					if($line[0][0] == ';'){
						continue;
					}

					$clubId = intVal($line[0]);
					$clubName = $line[1];
					$clubShortName = $line[2];
					$distance = intVal($line[3]);
					if($lastClubId != $clubId){
						$lastClubId = $clubId;
						$club = new Club();
						$club->setClub($clubId, $clubName, $clubShortName);
						$this->clubList[] = $club;
					}
					$latestIndex = count($this->clubList)-1;
					for($i=4; $i<count($line); $i++){
						$date = intval($line[$i]);
						$this->clubList[$latestIndex]->addDate($distance, $date);
						$this->addDateList($date);
						if($distance == 600){
							$date = $this->plusDate($date);
							$this->addDateList($date);
						}
						if($distance == 1000){
							$date = $this->plusDate($date);
							$this->addDateList($date);
							$date = $this->plusDate($date);
							$this->addDateList($date);
						}
					}
				}
				fclose($handle);
			}
			sort($this->dateList);
		}

		function addDateList($date){
			if(array_search($date, $this->dateList) === false){
				$this->dateList[] = $date;
			}
		}

		function plusDate($date){
			$today = mktime(0,0,0,intVal($date/100), $date%100, $this->year);
			$tomorrow = $today + 86400;
			return intVal(date("md",$tomorrow ));
		}

		function getYear(){
			return $this->year;
		}

		function getDateCount(){
			return count($this->dateList);
		}

		function getMonth($id){
			return intval($this->dateList[$id] / 100);
		}

		function getDay($id){
			return $this->dateList[$id] % 100;
		}

		function getWeekDay($id){
			$dateValue = strtotime($this->year*10000 + $this->getMonth($id)*100 + $this->getDay($id));
			$weekDayIndex = date("w", $dateValue);
			$weekDayArray = array("日", "月", "火", "水", "木", "金", "土");
			return $weekDayArray[$weekDayIndex];
		}

		function getBRMData($date){
			global $distanceArray;

			$BRMDatalist = array();
			for($i=0; $i<count($distanceArray)-1; $i++){
				foreach($this->clubList as &$club){
					$club->getBRMData($date, $i, $BRMDatalist);
				}
			}

			return $BRMDatalist;
		}


	};

	class BRMData {
		var $brmArray;

		// 初期化
		function initialize(){
			$this->brmArray = array();

			$year=START_YEAR;
			for(;;){
				if(file_exists($year . ".txt") == false){
					break;
				}
				$brm = new BRM();
				$brm->initialize($year);
				array_push($this->brmArray, $brm);
				$year++;
			}
		}

		function getYearCount(){
			return count($this->brmArray);
		}

		function getYear($id){
			return $this->brmArray[$id]->getYear();
		}

		function getDateCount($year){
			return $this->brmArray[$year-START_YEAR]->getDateCount();
		}

		function getMonth($year, $id){
			return $this->brmArray[$year-START_YEAR]->getMonth($id);
		}

		function getDay($year, $id){
			return $this->brmArray[$year-START_YEAR]->getDay($id);
		}

		function getWeekDay($year, $id){
			return $this->brmArray[$year-START_YEAR]->getWeekDay($id);
		}

		function getBRMData($year, $date){
			return $this->brmArray[$year-START_YEAR]->getBRMData($date);
		}
	}
?>


brmcalendar/index.php
<?php
	include 'brmdata.php';

	$brm = new BRMData();
	$brm->initialize();
	$yearCount = $brm->getYearCount();
	if($yearCount == 0){
		exit;
	}
	$year = $brm->getYear(0);
	$dayCount = $brm->getDateCount($year);
?>
<!doctype html>
<html manifest="cache.manifest">
<head>
<meta charset="utf-8">
<meta name="description" content="BRM Calendar">
<meta name="author" content="MM">
<meta name="format-detection" content="telephone=no">
<meta name="apple-mobile-web-app-capable" content="yes">
<meta name="apple-mobile-web-app-status-bar-style" content="default">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<link rel="apple-touch-icon" href="icon.png" />
<link rel="stylesheet" href="index.css">
<style type='text/css'>
<?php
	for($i=0; $i<$dayCount; $i++){
		$top = $i*26+6;
		echo "#pos$i {position:absolute; top:" . $top . "px; left:55px;}\n";
	}
?>
</style>
<script type="text/javascript">
	function logEvent(event) {
		console.log(event.type);
	}
	window.applicationCache.addEventListener('checking',logEvent,false);
	window.applicationCache.addEventListener('noupdate',logEvent,false);
	window.applicationCache.addEventListener('downloading',logEvent,false);
	window.applicationCache.addEventListener('cached',logEvent,false);
	window.applicationCache.addEventListener('updateready',logEvent,false);
	window.applicationCache.addEventListener('obsolete',logEvent,false);
	window.applicationCache.addEventListener('error',logEvent,false);
</script>
<title>BRM Calendar</title>
</head>
<body>
<div id="calendar">
<table>
<?php
	for($i=0; $i<$dayCount; $i++){
		$month = $brm->getMonth($year, $i);
		$day = $brm->getDay($year, $i);
		$weekDay = $brm->getWeekDay($year, $i);

		if($weekDay == "土"){
			echo "<tr class=\"sat\">\n";
		}
		else if($weekDay == "日"){
			echo "<tr class=\"holiday\">\n";
		}
		else{
			echo "<tr class=\"other\">\n";
		}
		echo "<td class=\"date\">$month</td>\n";
		echo "<td class=\"date\">$day</td>\n";
		echo "<td class=\"date\">$weekDay</td>\n";
		echo "<td class=\"brm\"></td>\n";
		echo "</tr>\n";
	}
?>
</table>
</div>
<?php
	$prevBRMData = array();
	for($i=0; $i<$dayCount; $i++){
		$month = $brm->getMonth($year, $i);
		$day = $brm->getDay($year, $i);

		echo "<div id=\"pos$i\">\n";
		$BRMData = $brm->getBRMData($year, $month*100+$day);
		for($j=0; $j<count($prevBRMData); $j++){
			if($prevBRMData[$j] > 0){
				$prevBRMData[$j]--;
			}
		}
		for($j=0, $k=0; $j<count($BRMData); $j++, $k++){
			while($prevBRMData[$k] > 0){
				echo "<span class=\"empty\"></span>\n";
				$k++;
			}
			foreach($distanceArray as &$distance){
				if($distance == 0){
					break;
				}
				if(strstr($BRMData[$j], "" . $distance)){
					echo "<span class=\"event d$distance\">$BRMData[$j]</span>\n";

					if($distance==1000){
						$prevBRMData[$k] += 3;
					}
					elseif($distance==600){
						$prevBRMData[$k] += 2;
					}
					else{
						$prevBRMData[$k] += 1;
					}
					break;
				}
			}
		}
		echo "</div>\n";
	}
?>
</body>
</html>
